src/components/legend/legend_item.tsx

```diff
@@ -106,19 +140,39 @@ export class LegendListItem extends React.Component<LegendItemProps> {
       selectedLegendItem == null ? false : selectedLegendItem.seriesIdentifier.key === seriesIdentifier.key;
     const hasTitleClickListener = Boolean(onLegendItemClickListener);
     const itemClasses = classNames('echLegendItem', `echLegendItem--${legendPosition}`, {
-      'echLegendItem-isHidden': !isSeriesVisible,
+      'echLegendItem--hidden': !isSeriesVisible,
       'echLegendItem__displayValue--hidden': !isLegendItemVisible,
     });
 
     return (
       <div className={itemClasses} onMouseEnter={this.onLegendItemMouseOver} onMouseLeave={this.onLegendItemMouseOut}>
-        {color && renderColor(color, isSeriesVisible)}
+        {color && this.renderColor(this.state.color ?? color, isSeriesVisible)}
         {label && renderTitle(label, onTitleClick, hasTitleClickListener, isSelected, showLegendDisplayValue)}
         {showLegendDisplayValue && renderDisplayValue(displayValue, isSeriesVisible)}
       </div>
     );
+  };
+
+  render() {
+    const { renderColorPicker } = this.props;
+    if (renderColorPicker && this.state.isOpen) {
+      return renderColorPicker(this.onColorChange, this.setToggleIsOpen, this.state.isOpen, this.renderItem());
+    }
+
+    return this.renderItem();
   }
 
+  setToggleIsOpen = () => {
+    this.setState({ isOpen: !this.state.isOpen });
```

Comment:
We should not reply on the current state object values (see https://reactjs.org/docs/state-and-lifecycle.html#state-updates-may-be-asynchronous)

```suggestion
    this.setState((state) => ({ isOpen: !state.isOpen }));
```